torch/_inductor/ir.py

```diff
@@ -5567,3 +5571,101 @@ def codegen_collective(self, wrapper, output_name, input_names):
             f"group={output_name}_pg, "
             "async_op=True)"
         )
+
+
+"""
+NOTE: `all_to_all_single` is implemented differently from other collectives
+because its output shape is data-dependent on input `output_split_sizes`.
+Instead of letting Inductor manage allocation of output buffer that has
+data-dependent dim size which is difficult to do, we resort to using
+`ExternKernelAlloc` and letting the underlying `fun_col_impl._all_to_all_single`
+kernel manage the output buffer allocation.
+"""
+
+
+class AllToAllSingle(ExternKernelAlloc):
+    def __init__(
+        self,
+        layout,
+        inputs,
+        constant_args,
+        inputs_allow_none,
+    ):
+        super().__init__(layout, inputs, constant_args)
+        self.inputs = inputs
+        self.inputs_allow_none = inputs_allow_none
```

Comment:
These two fields are kind of weird, not really precedent for it. (Maybe it's OK, I don't have an intuitive sense in IR for if it is ok or not)

Reply:
Updated to use `arg_index_to_input_index` dict instead, to hopefully make it more clear for posterity